File: DynamicAudioNormalizerJNI/samples/com/muldersoft/dynaudnorm/samples/DynamicAudioNormalizerExample.java
//////////////////////////////////////////////////////////////////////////////////
// Dynamic Audio Normalizer - Java/JNI Wrapper
// Copyright (c) 2014-2017 LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
// http://opensource.org/licenses/MIT
//////////////////////////////////////////////////////////////////////////////////

package com.muldersoft.dynaudnorm.samples;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.util.Map;

import com.muldersoft.dynaudnorm.JDynamicAudioNormalizer;

public class DynamicAudioNormalizerExample {

	//Default arguments
	private static final int CHANNELS_DEFAULT = 2;
	private static final int SAMPLE_RATE_DEFAULT = 44100;
	private static final AudioFileIO.PCMFileType AUDIO_FORMAT_DEFAULT = AudioFileIO.PCMFileType.PCM_16BIT_LE;
	
	//Normalizer options
	private static final int FRAME_LEN = 500;	//Frame length in MSec
	private static final int FILTER_SIZE = 31;	//Filter size in frames
	
	//------------------------------------------------------------------------------------------------
	// Print Logo
	//------------------------------------------------------------------------------------------------
	
	private static void printLogo() {
		int [] version = null;
		Map<String,String> buildInfo = null;
		try	{
			try	{
				version = JDynamicAudioNormalizer.getVersionInfo();
				buildInfo = JDynamicAudioNormalizer.getBuildInfo();
			}
			catch(Exception ex) {
				throw new Error("Failed to initialize DynamicAudioNormalizer library!", ex);
			}
		}
		finally {
			System.out.println("\n===========================================================================");
			if((version != null) && (version.length >= 3)) {
				System.out.printf("Dynamic Audio Normalizer, Version %d.%02d-%d\n", version[0], version[1], version[2]);
			}
			else {
				System.out.println("Dynamic Audio Normalizer, Unknown Version");
			}
			if((buildInfo != null) && buildInfo.containsKey("BuildDate")) {
				System.out.printf("Copyright (c) 2014-%s LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.\n", buildInfo.get("BuildDate").substring(7));
			}
			else {
				System.out.println("Copyright (c) 2014-2017 LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.");
			}
			System.out.println();
			System.out.println("This program is free software: you can redistribute it and/or modify");
			System.out.println("it under the terms of the GNU General Public License <http://www.gnu.org/>.");
			System.out.println("Note that this program is distributed with ABSOLUTELY NO WARRANTY.");
			System.out.println("===========================================================================\n");
		}
	}
	
	//------------------------------------------------------------------------------------------------
	// Processing Loop
	//------------------------------------------------------------------------------------------------

	public static void processFile(final String inputFile, final String outputFile, final int channles, final int sampleRate, final AudioFileIO.PCMFileType audioFormat) throws NoSuchAlgorithmException, IOException {
		// Open input and output files
		System.out.print("Opening input and out files...");
		try(AudioFileIO.PCMFileReader reader = new AudioFileIO.PCMFileReader(inputFile, channles, audioFormat)) {
			try (AudioFileIO.PCMFileWriter writer =  new AudioFileIO.PCMFileWriter(outputFile, channles, audioFormat)) {
				double[][] sampleBuffer = new double[channles][4096];
				
				System.out.print(" Done!\nCreating normalizer instance...");
				try(final JDynamicAudioNormalizer instance = new JDynamicAudioNormalizer(channles, sampleRate, FRAME_LEN, FILTER_SIZE, 0.95, 10.0, 0.0, 0.0, true, false, false)) {
					// Process samples
					System.out.print(" Done!\nProcessing input samples...");
					int progressUpdate = 0;
					for (;;) {
						int sampleCount = 0;
						try {
							sampleCount = reader.read(sampleBuffer);
						}
						catch (IOException e) {
							throw new IOException("Failed to read data from input file!", e);
						}
						
						if (sampleCount > 0) {
							final long outputSamples = instance.processInplace(sampleBuffer, sampleCount);
							if (outputSamples > 0) {
								try {
									writer.write(sampleBuffer, (int) outputSamples);
								}
								catch (IOException e) {
									throw new IOException("Failed to write data to output file!", e);
								}
							}
							if(++progressUpdate >= 13) {
								progressUpdate = 0;
								System.out.print(".");
							}
						}
						
						if (sampleCount < sampleBuffer[0].length) {
							break; /* EOF reached */
						}
					}
					
					// Flush pending samples
					System.out.print(" Done!\nFlushing remaining samples...");
					for (;;) {
						final long outputSamples = instance.flushBuffer(sampleBuffer);
						if (outputSamples > 0) {
							try {
								writer.write(sampleBuffer, (int) outputSamples);
							}
							catch (IOException e) {
								throw new IOException("Failed to write data to output file!", e);
							}
							if(++progressUpdate >= 13) {
								progressUpdate = 0;
								System.out.print(".");
							}
							
						}
						else {
							break; /* No more samples */
						}
					}
					
					// Reset
					System.out.print(" Done!\nResetting normalizer instance...");
					instance.reset(); /* This is NOT normally required, but we do it here to test the reset() API */
					
					// Close input and output
					System.out.print(" Done!\nShutting down..."); /*Implicitly, due to AutoCloseable*/
				}
			}
		}
		
		// Finished
		System.out.println(" Done!\n\nAll done. Goodbye.");
	}
	
	//------------------------------------------------------------------------------------------------
	// Main
	//------------------------------------------------------------------------------------------------

	public static void main(final String [] args)
	{
		try {
			//Print program logo
			printLogo();
			
			//Check number of arguments
			if(args.length < 2) {
				System.out.println("Usage:");
				System.out.println("  java -jar Example.jar [<channel_count> [<sample_rate> [<sample_format>]]] <input.pcm> <output.pcm>");
				System.exit(-1);
			}
			
			//Initialize arguments to default values
			int channels = CHANNELS_DEFAULT;
			int sampleRate = SAMPLE_RATE_DEFAULT;
			AudioFileIO.PCMFileType audioFormat = AUDIO_FORMAT_DEFAULT;
			
			//Parse optional arguments from command-line
			try {
				if(args.length > 2) {
					channels = Integer.parseInt(args[0]);
					if(args.length > 3) {
						sampleRate = Integer.parseInt(args[1]);
						if(args.length > 4) {
							audioFormat = AudioFileIO.PCMFileType.valueOf(args[2]);
						}
					}
				}
			}
			catch(Exception e) {
				throw new IllegalArgumentException("Invalid command-line argument given!", e);
			}
			
			//Process the audio file
			final int offset = Math.min(args.length, 5) - 2;
			processFile(args[offset], args[offset + 1], channels, sampleRate, audioFormat);
		}
		catch(Throwable e) {
			System.err.println("\n\nERROR: " + e.getMessage() + "\n");
			e.printStackTrace(System.err);
		}
	}

}

File: DynamicAudioNormalizerPYD/samples/DynamicAudioNormalizerExample.py
##################################################################################
# Dynamic Audio Normalizer - Python Wrapper
# Copyright (c) 2014-2017 LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# http://opensource.org/licenses/MIT
##################################################################################

#import stdlib modules
import sys
import os
import datetime
import array

#import PyDynamicAudioNormalizer
from DynamicAudioNormalizer import PyDynamicAudioNormalizer

#import Wave Reader and Writer
from WaveFileUtils import WaveFileReader
from WaveFileUtils import WaveFileWriter


#----------------------------------------------------------------------
# Utility Functions
#----------------------------------------------------------------------

def alloc_sample_buffers(channles, length):
	buffers = ()
	for i in range(0, channles):
		arr = array.array('d')
		for j in range(0, length):
			arr.append(0.0)
		buffers += (arr,)
	return buffers


#----------------------------------------------------------------------
# Logging Callback
#----------------------------------------------------------------------

logfile = None

def my_log_func(level, message):
	if logfile:
		logfile.write("[{}][{}] {}\n".format(datetime.datetime.utcnow().isoformat(), int(level), message))
		logfile.flush()


#----------------------------------------------------------------------
# Processing Loop
#----------------------------------------------------------------------

def normalize_samples(wav_reader, wav_writer, logfile, config):
	print(" Done!\nInitializing...", end = '', flush = True)
	if logfile:
		PyDynamicAudioNormalizer.setLogFunction(my_log_func)
	with PyDynamicAudioNormalizer(wav_reader.getChannels(), wav_reader.getSamplerate(), *config) as normalizer:
		indicator, buffersIn, buffersOut = 0, alloc_sample_buffers(wav_reader.getChannels(), 4096), alloc_sample_buffers(wav_reader.getChannels(), 4096)
		print(" Done!\nProcessing audio samples...", end = '', flush = True)
		while True:
			count = wav_reader.read(buffersIn)
			if count:
				count = normalizer.process(buffersIn, buffersOut, count)
				if count:
					wav_writer.write(buffersOut, count)
				indicator += 1
				if indicator >= 7:
					print('.', end = '', flush = True)
					indicator = 0
				continue
			break;
		print(" Done!\nFlushing buffers...", end = '', flush = True)
		while True:
			count = normalizer.flushBuffer(buffersOut)
			if count:
				wav_writer.write(buffersOut, count)
				indicator += 1
				if indicator >= 7:
					print('.', end = '', flush = True)
					indicator = 0
				continue
			break;
		print(' Done!\n')


#----------------------------------------------------------------------
# Main
#----------------------------------------------------------------------

version = PyDynamicAudioNormalizer.getVersion()
print("Dynamic Audio Normalizer, Version {}.{}-{} [{}]".format(*version[0], version[1][4]))
print("Copyright (c) 2014-{} LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.".format(version[1][0][7:]))
print("Built on {} at {} with {} for Python-{}.\n".format(*version[1]))

if (len(sys.argv) >= 5):
	config = (int(sys.argv[1]), int(sys.argv[2]))
	source_file = os.path.abspath(sys.argv[3])
	output_file = os.path.abspath(sys.argv[4])
elif (len(sys.argv) >= 4):
	config = (int(sys.argv[1]),)
	source_file = os.path.abspath(sys.argv[2])
	output_file = os.path.abspath(sys.argv[3])
elif (len(sys.argv) >= 3):
	config = ()
	source_file = os.path.abspath(sys.argv[1])
	output_file = os.path.abspath(sys.argv[2])
else:
	print("Usage:\n  {} {} [[<filterSize>] frameLen] <input.wav> <output.wav>\n".format(os.path.basename(sys.executable), os.path.basename(__file__)), file=sys.stderr)
	sys.exit(1)

if not os.path.isfile(source_file):
	print("Input file \"{}\" not found!\n".format(source_file), file=sys.stderr)
	sys.exit(1)

print("Source file: \"{}\"".format(source_file))
print("Output file: \"{}\"".format(output_file))

logfilePath = os.environ.get('LOGFILE') 
if logfilePath:
	print("Report file: \"{}\"".format(logfilePath))
	logfile = open(logfilePath, 'w')
else:
	logfile = None

print("\nOpening input and output files...", end = '', flush = True)
with WaveFileReader(source_file) as wav_reader:
	with WaveFileWriter(output_file, wav_reader.getChannels(), wav_reader.getSampleWidth(), wav_reader.getSamplerate()) as wav_writer:
		normalize_samples(wav_reader, wav_writer, logfile, config)
		print('All is done. Goodbye.')

if logfile:
	logfile.close()
